core/src/test/scala/org/apache/spark/CacheRecoveryManagerSuite.scala

```diff
@@ -0,0 +1,200 @@
+import org.apache.spark.storage.{BlockId, BlockManagerId, RDDBlockId}
+import org.apache.spark.storage.BlockManagerMessages._
+import org.apache.spark.util.ThreadUtils
+
+class CacheRecoveryManagerSuite
+  extends SparkFunSuite with MockitoSugar with Matchers with Eventually {
+
+  val oneGB: Long = 1024L * 1024L * 1024L * 1024L
+  val plentyOfMem = Map(
+    BlockManagerId("1", "host", 12, None) -> ((oneGB, oneGB)),
+    BlockManagerId("2", "host", 12, None) -> ((oneGB, oneGB)),
+    BlockManagerId("3", "host", 12, None) -> ((oneGB, oneGB)))
+
+  test("CacheRecoveryManager will replicate blocks until empty and then kill executor") {
+    val conf = new SparkConf()
+    val eam = mock[ExecutorAllocationManager]
+    val blocks = Seq(RDDBlockId(1, 1), RDDBlockId(2, 1))
+    val bmme = FakeBMM(1, blocks.iterator, plentyOfMem)
+    val bmmeRef = DummyRef(bmme)
+    val cacheRecoveryManager = new CacheRecoveryManager(bmmeRef, eam, conf)
+    when(eam.killExecutors(Seq("1"))).thenReturn(Seq("1"))
+    val result = cacheRecoveryManager.startCacheRecovery(Seq("1"))
+
+    eventually {
+      verify(eam).killExecutors(Seq("1"))
+      bmme.replicated.get("1").get shouldBe 2
+    }
+
+    cleanup(result, cacheRecoveryManager)
+  }
+
+  test("CacheRecoveryManager will kill executor if it takes too long to replicate") {
+    val conf = new SparkConf().set(DYN_ALLOCATION_CACHE_RECOVERY_TIMEOUT.key, "1s")
+    val eam = mock[ExecutorAllocationManager]
+    val blocks = Set(RDDBlockId(1, 1), RDDBlockId(2, 1), RDDBlockId(3, 1), RDDBlockId(4, 1))
+    val bmme = FakeBMM(600, blocks.iterator, plentyOfMem)
+    val bmmeRef = DummyRef(bmme)
+    val cacheRecoveryManager = new CacheRecoveryManager(bmmeRef, eam, conf)
+    val result = cacheRecoveryManager.startCacheRecovery(Seq("1"))
+
+    eventually(timeout(Span(1010, Millis)), interval(Span(500, Millis))) {
+      verify(eam, times(1)).killExecutors(Seq("1"))
+      bmme.replicated.get("1").get shouldBe 1
+    }
+
+    cleanup(result, cacheRecoveryManager)
+  }
+
+  test("shutdown timer will get cancelled if replication finishes") {
+    val conf = new SparkConf().set(DYN_ALLOCATION_CACHE_RECOVERY_TIMEOUT.key, "1s")
+    val eam = mock[ExecutorAllocationManager]
+    val blocks = Set(RDDBlockId(1, 1))
+    val bmme = FakeBMM(1, blocks.iterator, plentyOfMem)
+    val bmmeRef = DummyRef(bmme)
+    val cacheRecoveryManager = new CacheRecoveryManager(bmmeRef, eam, conf)
+
+    val result = cacheRecoveryManager.startCacheRecovery(Seq("1"))
+    val minimumTime = System.currentTimeMillis() + 1000
+
+    eventually(timeout(Span(1500, Millis)), interval(Span(500, Millis))) {
+      // should be killed once not twice
+      verify(eam, times(1)).killExecutors(Seq("1"))
+      // wait at least a second to be sure we don't kill executor twice
+      System.currentTimeMillis() should be > minimumTime
+    }
+
+    cleanup(result, cacheRecoveryManager)
+  }
+
+  test("Blocks won't replicate if we are running out of space") {
```

Comment:
nit: start test names with lower case (looks slightly better in output)

Reply:
fixed.